include/zephyr/sensing/sensing.h

```diff
@@ -0,0 +1,352 @@
+
+	/** Sensor index in the sensor info array returned from sensing_get_sensors() */
+	int32_t sensor_index;
+
+	/** Sensor flags */
+	uint32_t flags;
+
+	/** Minimal report interval in micro seconds */
+	uint32_t minimal_interval;
+
+	/** Sensor version */
+	struct sensing_sensor_version version;
+};
+
+/**
+ * @struct sensing_callback_list
+ * @brief Sensor subsystem event callback list
+ *
+ */
+struct sensing_callback_list {
+	sensing_data_event_t on_data_event;
+	sensing_bias_event_t on_bios_event;
+	sensing_flush_complete_t on_flush_complete;
+};
+
+/**
+ * @struct sensing_sensor_config
+ * @brief Sensor subsystem sensor configure, including interval, sensitivity, latency
+ *
+ */
+struct sensing_sensor_config {
+	enum sensing_sensor_attribute attri;
+	int8_t data_field;
+	union {
+		uint32_t interval;
+		uint32_t sensitivity;
+		uint64_t latency;
+	};
+};
+
+
+ /**
+  * @brief Get all supported sensor instances' information.
+  *
+  * This API just returns read only information of sensor instances,
+  * no side effect to sensor instances.
+  *
+  * @param num_sensors Get number of sensor instances.
+  *
+  * @param info For receiving sensor instances' information array pointer.
+  *
+  * @return 0 on success or negative error value on failure.
+  */
+int sensing_get_sensors(int *num_sensors, const struct sensing_sensor_info **info);
+
+/**
+ * @brief Open sensor instance by index.
+ *
+ * Application clients use it to open a sensor instance and get its handle.
+ * Support multiple Application clients for open same sensor instance,
+ * in this case, the returned handle will different for different clients.
+ * meanwhile, also register sensing callback list.
+ *
+ * @param index The sensor index in sensor info array which get from /ref int sensing_get_sensors().
+ *
+ * @param cb_list callback list to be registered to sensing.
+ *
+ * @param *handle The opened instance handle, if failed will be set to -1.
+ *
+ * @return 0 on success or negative error value on failure.
+ */
+int sensing_open_sensor(
+		int index, const struct sensing_callback_list *cb_list,
+		int *handle);
+
+/**
+ * @brief Open sensor instance by sensor type and sensor index.
+ *
+ * Application clients use it to open a sensor instance and get its handle.
+ * Support multiple Application clients for open same sensor instance,
+ * in this case, the returned handle will different for different clients.
+ * meanwhile, also register sensing callback list
+ *
+ * @param type The sensor type which need to open.
+ *
+ * @param instance The id for multiple instances with same sensor /ref type, 0 means the default
+ *                 instance for the /ref type.
+ *
+ * @param cb_list callback list to be registered to sensing.
+ *
+ * @param *handle The opened instance handle, if failed will be set to -1.
+ *
+ * @return 0 on success or negative error value on failure.
+ */
+int sensing_open_sensor_by_type(
+		int type, int instance, const struct sensing_callback_list *cb_list,
```

Comment:
If every sensor instances including physical sensor and virtual sensor clearly defined in devicetree, and with chosen, can map different sensor instance in devicetree, I think this is a good idea. 

So, we can change the concept of `type` + `sensor_index/instance` to identify a sensor by just using devicetree node, it's more simple and exactly mapped to devicetree. 

I had long discussions with @yperess before on `sensor_index/instance`, thank you @keith-zephyr , it's really a good idea! Will change to it.

Reply:
> Using `instance` here will be error prone for users. There's no checking/verification that the instance is unique. Any changes to the devictree instance also means the user must change the application C code to match.
> 
> It would be much cleaner for the application retrieve the device reference directly from the devicetree using something like:
> 
> ```
> sensing_open_sensor(DEVICE_DT_GET(DT_NODELABEL(base_accel)), cb_list, handle);
> sensing_open_sensor(DEVICE_DT_GET(DT_CHOSEN(zephyr_sensing_base_accel), cb_list, handle);
> ```

I like this as well, good idea!